src/index/index.c

```diff
@@ -0,0 +1,107 @@
+#include "index.h"
+
+int compareNodes(const void *p1, const void *p2) {
+  Node *a = (Node *)p1;
+  Node *b = (Node *)p2;
+
+  if (a->id > b->id) {
+    return 1;
+  } else if (a->id < b->id) {
+    return -1;
+  } else {
+    return 0;
+  }
+}
+
+int compareSI(void *p1, void *p2, void *ctx) {
+  SIValue *a = p1, *b = p2;
+
+  if (a->type & b->type & T_STRING) {
+    return strcmp(a->stringval, b->stringval);
+  } else if ((a->type & SI_NUMERIC) && (b->type & SI_NUMERIC)) {
+
+    if (a->doubleval > b->doubleval) {
+      return 1;
+    } else if (a->doubleval < b->doubleval) {
+      return -1;
+    } else {
+      return 0;
+    }
+  }
+
+  // We can only compare string and numeric SIValues, so this point
+  // should be unreachable for a properly constructed index.
+
+  return 0;
+}
+
+void freeVal(void *p1) {
+  free(p1);
+}
+
+skiplist* createIndex(RedisModuleCtx *ctx, const char *graphName, AST_IndexOpNode *indexOp) {
+  const char *label = indexOp->target.label;
+  const char *index_prop = indexOp->target.property;
+
+  LabelStore *store = LabelStore_Get(ctx, STORE_NODE, graphName, label);
+
+  LabelStoreIterator *it_test = LabelStore_Search(store, "");
+
+  skiplist *sl = skiplistCreate(compareSI, NULL, compareNodes, freeVal);
+
+  char *nodeId;
+  tm_len_t nodeIdLen;
+  Node *node;
+  EntityProperty *prop;
+
+  SIType lastKeyType = SI_NUMERIC | T_STRING;
+  SIType xor;
+
+  while(LabelStoreIterator_Next(it_test, &nodeId, &nodeIdLen, (void**)&node)) {
+    for (int i = 0; i < node->prop_count; i ++) {
+      prop = node->properties + i;
+
+      if (!strcmp(index_prop, prop->name)) {
+        SIValue *key = malloc(sizeof(SIValue));
```

Comment:
@swilly22 

There were a few interlinked flaws related to this area. I have made a few changes in this PR to amend them:

First, I have renamed 'obj' to 'key' in all skiplist code, because I was confusing myself and making mistakes because I was forgetting whether 'obj' referred to the key or the value.

In my initial updates to the skiplist (already in master), I added a function pointer to the skiplist struct so that the skiplist could free keys, but there was no corresponding create routine. This has now been added (as an example, see 'cloneKey' in index.c).

This is important because the skiplist/index must own the SIValues that they use as keys - otherwise, they would be pointers to the properties on the GraphEntity objects themselves, which are volatile as they can be updated through separate queries.

TODO - functions that update or delete GraphEntity properties referenced by an index must make corresponding updates to the index; this is not yet implemented.

Custom clone routines are necessary because
A) the skiplist has no knowledge of the type of its keys, and when we introduce support for composite indices, cloning will be a more complex operation.
B) We cannot allocate a key outside of the skiplistInsert routine because we don't have knowledge of whether we're appending a value to an existing key or inserting a new key.